File: backend/league/models/budget.py
from decimal import Decimal
from django.db import models


BUDGET_DIVISIONS = [
    # Baseball
    ("TEE_BALL",   "Tee Ball"),
    ("PEE_WEE",    "Pee Wee"),
    ("AAA",        "AAA"),
    ("MAJORS",     "Majors"),
    # Softball
    ("SF_MINORS",  "Softball Minors"),
    ("SF_MAJORS",  "Softball Majors"),
    # Cross-sport
    ("ALL_DIV",    "All Divisions"),
]

BUDGET_CATEGORIES = [
    ("BASEBALL",     "Baseball"),
    ("SOFTBALL",     "Softball"),
    ("CONCESSIONS",  "Concessions"),
    ("MARKETING",    "Marketing"),
    ("GROUNDS",      "Grounds & Facilities"),
    ("RENT_UTIL",    "Rent & Utilities"),
    ("EQUIPMENT",    "Equipment"),
    ("ADMIN",        "Admin & Operations"),
    ("SPONSORSHIP",  "Sponsorship & Fundraising"),
    ("LL_FEES",      "Little League Fees"),
    ("SAFETY",       "Safety & Supplies"),
    ("APPAREL",      "Apparel"),
    ("SCHOLARSHIPS", "Scholarships"),
    ("DONATIONS",    "Donations"),
    ("OTHER",        "Other"),
]


class BudgetLine(models.Model):
    """A single budget line item for a given year."""

    year = models.PositiveIntegerField(db_index=True)
    category = models.CharField(max_length=20, choices=BUDGET_CATEGORIES)
    item = models.CharField(max_length=200)
    owner_role = models.CharField(max_length=150, blank=True)
    is_revenue = models.BooleanField(default=False)

    actual = models.DecimalField(
        max_digits=10, decimal_places=2, null=True, blank=True,
        verbose_name="Prior year actual ($)",
    )
    estimate = models.DecimalField(
        max_digits=10, decimal_places=2, null=True, blank=True,
        verbose_name="Current year estimate ($)",
    )
    # When True the estimate was hand-entered; when False it auto-follows actual * 1.05
    estimate_override = models.BooleanField(default=False)

    # Quantity-based pricing — when both are set, total = quantity × unit_cost
    quantity = models.PositiveIntegerField(
        null=True, blank=True,
        help_text="Number of units (e.g. 44 teams × $10 = $440).",
    )
    unit_cost = models.DecimalField(
        max_digits=10, decimal_places=2, null=True, blank=True,
        help_text="Cost per unit. Total = quantity × unit_cost.",
    )

    division = models.CharField(
        max_length=20, blank=True, choices=BUDGET_DIVISIONS,
        help_text="Baseball/Softball division this item applies to. Leave blank for cross-division items.",
    )

    sub_group = models.CharField(
        max_length=100, blank=True,
        help_text="Optional grouping within a category (e.g. 'Regular Season', 'Winter Workout')",
    )
    notes = models.TextField(blank=True)
    sort_order = models.PositiveIntegerField(default=0)

    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    class Meta:
        ordering = ["year", "category", "sort_order", "id"]

    def __str__(self):
        return f"{self.year} | {self.category} | {self.item}"

    @property
    def effective_estimate(self):
        """
        The estimate to display, in priority order:
        1. quantity × unit_cost  (auto-calculated if both are set)
        2. hand-entered estimate_override
        3. actual × 1.05
        4. estimate (may be None)
        """
        if self.quantity is not None and self.unit_cost is not None:
            return (Decimal(self.quantity) * self.unit_cost).quantize(Decimal("0.01"))
        if self.estimate_override and self.estimate is not None:
            return self.estimate
        if self.actual is not None:
            return (self.actual * Decimal("1.05")).quantize(Decimal("0.01"))
        return self.estimate  # may be None


class BudgetApproval(models.Model):
    """Records a Treasurer's approval of the budget for a given year."""

    year = models.PositiveIntegerField(unique=True)
    approved_by = models.CharField(max_length=150)
    approved_at = models.DateTimeField(auto_now_add=True)
    notes = models.TextField(blank=True)

    def __str__(self):
        return f"{self.year} approved by {self.approved_by}"

File: frontend/src/api/budget.ts
import client from "./client"

/** Divisions that apply to Baseball / Softball budget items. */
export const DIVISIONS = [
  { value: "TEE_BALL",  label: "Tee Ball",         sport: "baseball", color: "#78909c" },
  { value: "PEE_WEE",  label: "Pee Wee",           sport: "baseball", color: "#26a69a" },
  { value: "AAA",       label: "AAA",               sport: "baseball", color: "#1565c0" },
  { value: "MAJORS",    label: "Majors",            sport: "baseball", color: "#C41230" },
  { value: "SF_MINORS", label: "Softball Minors",   sport: "softball", color: "#8e24aa" },
  { value: "SF_MAJORS", label: "Softball Majors",   sport: "softball", color: "#d81b60" },
  { value: "ALL_DIV",   label: "All Divisions",     sport: "both",     color: "#546e7a" },
]

/** Categories that support per-division line items. */
export const DIVISION_CATEGORIES = new Set(["BASEBALL", "SOFTBALL"])

export const CATEGORIES = [
  { value: "BASEBALL",     label: "Baseball",                 color: "#1565c0" },
  { value: "SOFTBALL",     label: "Softball",                 color: "#6a1b9a" },
  { value: "CONCESSIONS",  label: "Concessions",              color: "#e65100" },
  { value: "MARKETING",    label: "Marketing",                color: "#00695c" },
  { value: "GROUNDS",      label: "Grounds & Facilities",     color: "#2e7d32" },
  { value: "RENT_UTIL",    label: "Rent & Utilities",         color: "#4527a0" },
  { value: "EQUIPMENT",    label: "Equipment",                color: "#1565c0" },
  { value: "ADMIN",        label: "Admin & Operations",       color: "#37474f" },
  { value: "SPONSORSHIP",  label: "Sponsorship & Fundraising",color: "#d97706" },
  { value: "LL_FEES",      label: "Little League Fees",       color: "#c62828" },
  { value: "SAFETY",       label: "Safety & Supplies",        color: "#e65100" },
  { value: "APPAREL",      label: "Apparel",                  color: "#00695c" },
  { value: "SCHOLARSHIPS", label: "Scholarships",             color: "#4a148c" },
  { value: "DONATIONS",    label: "Donations",                color: "#1b5e20" },
  { value: "OTHER",        label: "Other",                    color: "#546e7a" },
]

export interface BudgetLine {
  id: number
  year: number
  category: string
  category_display: string
  item: string
  sub_group: string
  owner_role: string
  is_revenue: boolean
  division: string            // e.g. "MAJORS", "AAA", "" for none
  actual: string | null       // Decimal comes as string from DRF
  estimate: string | null
  estimate_override: boolean
  quantity: number | null
  unit_cost: string | null    // Decimal from DRF
  effective_estimate: string | null
  notes: string
  sort_order: number
  created_at: string
  updated_at: string
}

export interface BudgetSummary {
  year: string | null
  by_category: Record<string, { label: string; actual: number; estimate: number }>
  total_expenses: { actual: number; estimate: number }
  total_revenue: { actual: number; estimate: number }
  gap: { actual: number; estimate: number }
  approval: BudgetApproval | null
}

export interface BudgetApproval {
  id: number
  year: number
  approved_by: string
  approved_at: string
  notes: string
}

function unwrap<T>(res: any): T {
  return res?.data?.data ?? res?.data ?? res
}

export async function getBudgetLines(year?: number): Promise<BudgetLine[]> {
  const res = await client.get("/budget/lines/", { params: year ? { year } : {} })
  return unwrap<BudgetLine[]>(res) ?? []
}

export async function createBudgetLine(data: Partial<BudgetLine>): Promise<BudgetLine> {
  const res = await client.post("/budget/lines/", data)
  return unwrap<BudgetLine>(res)
}

export async function updateBudgetLine(id: number, data: Partial<BudgetLine>): Promise<BudgetLine> {
  const res = await client.patch(`/budget/lines/${id}/`, data)
  return unwrap<BudgetLine>(res)
}

export async function deleteBudgetLine(id: number): Promise<void> {
  await client.delete(`/budget/lines/${id}/`)
}

export async function getBudgetSummary(year?: number): Promise<BudgetSummary> {
  const res = await client.get("/budget/summary/", { params: year ? { year } : {} })
  return unwrap<BudgetSummary>(res)
}

export async function approveBudget(year: number, approvedBy: string, notes: string): Promise<BudgetApproval> {
  const res = await client.post("/budget/approve/", { year, approved_by: approvedBy, notes })
  return unwrap<BudgetApproval>(res)
}

export async function revokeApproval(year: number): Promise<void> {
  await client.delete(`/budget/approve/${year}/`)
}

export async function getBudgetYears(): Promise<number[]> {
  const res = await client.get("/budget/years/")
  return res.data.years ?? []
}

export async function copyBudgetYear(
  fromYear: number,
  toYear: number,
  multiplier = 1.05,
  overwrite = false,
): Promise<{ created: number; to_year: number; from_year: number }> {
  const res = await client.post("/budget/copy-year/", {
    from_year: fromYear, to_year: toYear, multiplier, overwrite,
  })
  return res.data
}

export function getBudgetExportURL(year?: number): string {
  const base = (client.defaults.baseURL ?? "").replace(/\/$/, "")
  return `${base}/budget/export/${year ? `?year=${year}` : ""}`
}
